### PR TITLE
Fix torch.cat()  performance regression on single core CPU

### PR DESCRIPTION
**Pull request by mingfeima**
This PR addresses the performance regression on `torch.cat()` on CPU with single thread.
Previous optimization #30806 introduced regression for several cases on pytorch operator benchmark.
See #33334 for detail.

### CODE REVIEW

**Comment by mingfeima** [comment]
So #30806 and #33534 together will make sure we have better performance for every single case in `cat_test.py` from [PT OP Benchmark](https://github.com/pytorch/pytorch/tree/master/benchmarks/operator_benchmark) compared with old TH implementation.

Notably, for two cases `cat_M128_N1024_K1_dim2` and `cat_M128_N128_K1_dim2`, #33534 is not as good as #30806, but still provide significant boost v.s. TH cat.

Abstract the result from @hl475 in the following table:

Test   Name | TH | 30806 | 33534 | 33534 v.s. TH
-- | -- | -- | -- | --
cat_M128_N128_K1_dim2_cpu | 1432.811 | 39.233 | 165.406 | 8.66
cat_M128_N1024_K1_dim2_cpu | 11414.43 | 256.177 | 1261.451 | 9.05



[This](https://github.com/pytorch/pytorch/issues/33334#issuecomment-588008262) is the result on my test machine, Intel(R) Xeon(R) Gold 5220 CPU, 2 * 18 cores @2.2GHz. You may take as a reference.

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -128,6 +138,14 @@ Tensor & _cat_out_cpu(Tensor& result, TensorList tensors, int64_t dim) {
   result_size[dim] = cat_dim_size;
   result.resize_(result_size);
 
+  // fast path for single thread when inputs are contiguous and not empty
+  bool use_serial_kernel = result.numel() < at::internal::GRAIN_SIZE || at::get_num_threads() == 1;
+  ScalarType dtype = notSkippedTensor.scalar_type();
+  if (use_serial_kernel && allContiguous && (dtype == ScalarType::Double || dtype == ScalarType::Float)) {
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/TensorShape.cpp]: result should be contiguous, or you will have incorrect data here.

```python
x = torch.randn((4, 3, 8, 8 ))                                                                                                                                                                                                        
y = torch.randn(x.shape)                                                                                                                                                                                                              
z = torch.cat((x,y))                                                                                                                                                                                                                  
w = z.clone().contiguous(memory_format=torch.channels_last)                                                                                                                                                                           
w.stride()  # (192, 1, 24, 3)                                                                                                                                                                                                                      
u = torch.cat((x,y),out=w)
(abs(w-z)).max() # >0!
```

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Incorrect if result is not contiguous

File: aten/src/ATen/native/TensorShape.cpp (hunk 1 above)
Comment (mingfeima) [reply aten/src/ATen/native/TensorShape.cpp]: OK, i see. Do we need add new test cases to cover this?

Comment (VitalyFedyunin) [reply aten/src/ATen/native/TensorShape.cpp]: Yes please.

Comment (mingfeima) [reply aten/src/ATen/native/TensorShape.cpp]: Fixed. Test case added.

**Review by VitalyFedyunin — APPROVED** [review]

File: aten/src/ATen/native/cpu/CatKernel.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      scalar_t* input_ptr = (scalar_t*)(inputs[j].data_ptr) + i * local_inner;
+      if (local_inner < Vec::size()) {
+        #ifndef _MSC_VER
+        # pragma unroll
+        #endif
+        for (int64_t k = 0; k < local_inner; k++) {
+          result_ptr[k] = input_ptr[k];
+        }
+      } else {
+        vec256::map(
```
Comment (maratsubkhankulov) [inline aten/src/ATen/native/cpu/CatKernel.cpp:54]: @mingfeima does this produce avx move instructions? Can you comment on why memcpy wasn't used here instead? I couldn't see the reasoning in the comments.

Comment (mingfeima) [reply aten/src/ATen/native/cpu/CatKernel.cpp:54]: `std::memcpy` is upto SSE (128b). And vec256::map will compile AVX and AVX2 assembly (pytorch choose AVX or AVX2 according to your cpu ability at runtime). Unfortunately, neither would generate AVX512 code... Anyway, for memory bandwidth bond cases, should not make too much difference.

**Pull request closed without merging**